mesa_frames/concrete/pandas/space.py

```diff
@@ -0,0 +1,171 @@
+from collections.abc import Callable, Sequence
+
+import numpy as np
+import pandas as pd
+from typing import Literal
+
+from mesa_frames.abstract.space import GridDF
+from mesa_frames.concrete.pandas.mixin import PandasMixin
+
+
+class GridPandas(GridDF, PandasMixin):
+    _agents: pd.DataFrame
+    _copy_with_method: dict[str, tuple[str, list[str]]] = {
+        "_agents": ("copy", ["deep"]),
+        "_cells": ("copy", ["deep"]),
+        "_cells_capacity": ("copy", []),
+        "_offsets": ("copy", ["deep"]),
+    }
+    _cells: pd.DataFrame
+    _cells_capacity: np.ndarray
+    _offsets: pd.DataFrame
+
+    def _empty_cell_condition(self, cap: np.ndarray) -> np.ndarray:
+        # Create a boolean mask of the same shape as cap
+        empty_mask = np.ones_like(cap, dtype=bool)
+
+        if not self._agents.empty:
+            # Get the coordinates of all agents
+            agent_coords = self._agents[self._pos_col_names].to_numpy(int)
+
+            # Mark cells containing agents as not empty
+            empty_mask[tuple(agent_coords.T)] = False
+
+        return empty_mask
+
+    def _generate_empty_grid(
+        self, dimensions: Sequence[int], capacity: int
+    ) -> np.ndarray:
+        if not capacity:
+            capacity = np.inf
+        return np.full(dimensions, capacity)
+
+    def _sample_cells(
+        self,
+        n: int | None,
+        with_replacement: bool,
+        condition: Callable[[np.ndarray], np.ndarray],
+        seed: int | None = None,
+        respect_capacity: bool = True,
+    ) -> pd.DataFrame:
+        # Set up the random number generator
+        if seed is None:
```

Comment:
Shouldn't the seeding be handled at model layer? Are there other places where this seeding pattern appears?

Reply:
I will merge for now before this PR gets even bigger.

Reply:
if seed is None, it uses the rng from the model, so that if you specified the seed for the model at the start, you will always get the same sampling operations when you run the model.  
otherwise if seed is specified, it uses the custom seed specified by the user. 
It gives more flexibility to the user, but as you point out the best practice would be to handle rng at the model level.
Should we remove the seed parameter entirely?

Reply:
Additionally, I find it more confusing to have the seed being specified in a decentralized way. Also not sure if there is ever a situation where user wants to keep everything the same but tweaks the sampling, for exploration purpose. Should the need happen, they can just sample several times and drop the early ones. So, yes, let's remove it.

Reply:
You're right, done.